src/melee/it/types.h

```diff
@@ -0,0 +1,854 @@
+    s32 xC94;                    // 0xc94
+    s32 xC98;                    // 0xc98
+    s32 xC9C;                    // Total damage taken?
+    s32 xCA0;                    // Last amount of damage taken?
+    s32 xCA4;                    // 0xca4
+    s32 xCA8;                    // 0xca8, hitlag related
+    s32 xCAC_angle;              // 0xcac
+    s32 xCB0_source_ply;         // 0xcb0, damage source ply number
+    s32 xCB4;                    // 0xcb4
+    f32 xCB8_outDamageDirection; // 0xcb8, updated @ 80078184
+    f32 xCBC_hitlagFrames;       // 0xcbc, hitlag frames remaining
+    f32 xCC0;                    // 0xcc0
+    s32 xCC4;                    // 0xcc4
+    f32 xCC8_knockback;          // 0xcc8
+    f32 xCCC_incDamageDirection; // Direction from which damage was applied?
+    f32 xCD0;                    // 0xcd0
+    f32 xCD4;                    // 0xcd4
+    f32 xCD8;                    // 0xcd8
+    f32 xCDC;                    // 0xcdc
+    f32 xCE0;                    // 0xce0
+    f32 xCE4;                    // 0xce4
+    f32 xCE8;                    // 0xce8
+    HSD_GObj* xCEC_fighterGObj;  // 0xcec
+    HSD_GObj* xCF0_itemGObj;     // 0xcf0, is a fp GObj, but is the owner of the
+
+    HSD_GObj* xCF4_fighterGObjUnk;
+
+    /// @at{CF8} @sz{4}
+    /// @brief The entity that was detected by this item's inert hitbox.
+    HSD_GObj* toucher;
+
+    u32 xCFC;
+
+    /// @at{D00} @sz{4}
+    /// @brief The entity that got grabbed by this item.
+    HSD_GObj* grab_victim;
+
+    /// @at{D04} @sz{4}
+    /// @brief The entity that collided with this item's hitbox?
+    HSD_GObj* atk_victim;
+
+    u8 xD08;
+    u8 xD09;
+    u8 xD0A;
+    u8 xD0B;
+
+    /// @at{D0C} @sz{4}
+    enum_t xD0C;
+
+    s32 xD10;
+
+    /// @at{D14} @sz{4}
+    HSD_GObjPredicate if_anim;
+
+    /// @at{D18} @sz{4}
+    HSD_GObjEvent on_phys;
+
+    /// @at{D1C} @sz{4}
+    HSD_GObjPredicate if_coll;
+
+    /// @at{D20} @sz{4}
+    /// @todo What does this mean?
+    HSD_GObjEvent on_accessory;
+
+    /// @at{D24} @sz{4}
+    /// @brief Runs when an entity is detected by this item's inert hibox.
+    HSD_GObjPredicate if_touched;
+
+    /// @at{D28} @sz{4}
+    /// @brief Runs after applying hitlag in damage.
+    /// @todo What function is @c 8026a62c?
+    HSD_GObjEvent on_enter_hitlag;
+
+    // 0xd2c, runs after exiting hitlag in hitlag, update proc 8026a200
+    HSD_GObjEvent on_exit_hitlag;
+
+    /// @at{D28} @sz{4}
+    /// @brief Runs when the item is jumped on.
+    /// @todo What function is @c 80269bac?
+    HSD_GObjPredicate if_jumped_on;
+
+    /// @at{D34} @sz{4}
+    /// @brief When grabbing a fighter, run this function on self.
+    HSD_GObjEvent on_grab;
+
+    /** @at{D38} @sz{4}
+     * @brief When grabbing a fighter, run this function on them.
+     *
+     * @p gobj0 - The victim of the grab. \n
+     * @p gobj1 - This item's entity.
+     */
+    HSD_GObjInteraction on_grabbed_for_victim;
```

Comment:
Started reworking callback names. `if_` means a predicate (returns a `BOOL`) while `on_` is an event (returns `void`).

Reply:
Don't like the `if_` predicate because we'll end up seeing `if (if_jumped_on()` which is weird. How about `is_jumped_on`, etc

Reply:
You don't think that's too suggestive of a boolean field?

Reply:
Or maybe just skip the prefix so we have `if (jumped_on`, `if (collided`, etc. It'll be pretty obvious from context when something is returning a bool.

Reply:
And drop the `on_` too?

Reply:
> Don't like the `if_` predicate because we'll end up seeing `if (if_jumped_on()` which is weird. How about `is_jumped_on`, etc

Keep in mind that can't happen exactly because it would have to at least be
```c
if (ip->if_jumped_on(fp))
```

Reply:
I like the `on_` prefix. `on_grab`/`on_anim`/`on_phys` are very clear what they're doing, but without the prefix `grab`/`anim`/`phys` could easily be misinterpreted as a number of other imperative things such as `do_grab`, `update_anim`, `recompute_phys` etc

Reply:
What do you propose for the collision predicate?

Reply:
How about just `collided`? Alternatively `did_collide` or `detected_collision`

Reply:
Ok so no prefixes to indicate a callback. That works for me, it's how Microsoft does it (`collided` for after a collision happens, `colliding` for before a collision happens).

Reply:
Before|After
-|-
`if_anim`|`animated`
`on_phys`|`physics_updated`
`if_coll`|`collided`
`on_accessory`|`on_accessory`
`if_touched`|`touched`
`on_enter_hitlag`|`entered_hitlag`
`on_exit_hitlag`|`exited_hitlag`
`if_jumped_on`|`jumped_on`
`on_grab`|`grab_dealt`
`on_grabbed_for_victim`|`grabbed_for_victim`

Reply:
I like it! Also will carry over to nice descriptive function names, like `void itCapsulePhysicsUpdated(void)`.